### PR TITLE
[kafka-clusters] introduce new integration

### PR DESCRIPTION
**Pull request by maorfr**
part of https://issues.redhat.com/browse/APPSRE-2727

as we are onboarding Managed Services API, we need to prepare to consume their API.
this integration will create Kafka clusters via OCM and will output a Secret with required information to namespaces which will be defined as consuming a Kafka cluster.

we currently only output `bootstrapServerHost`, as the API is still WIP.

### CODE REVIEW

File: reconcile/kafka_clusters.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        desired_state.append(item)
+    return desired_state
+
+
+@defer
+def run(dry_run, thread_pool_size=10,
+        internal=None, use_jump_host=True, defer=None):
+    settings = queries.get_app_interface_settings()
+    kafka_clusters = queries.get_kafka_clusters()
+    ocm_map = OCMMap(clusters=kafka_clusters,
```
Comment (apahim) [inline reconcile/kafka_clusters.py:50]: this fails while we don't define the first kafka cluster on app-interface

File: reconcile/kafka_clusters.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                namespace_info['name'],
+                resource.kind,
+                resource.name,
+                resource
+            )
+
+    ob.realize_data(dry_run, oc_map, ri)
+
+    if error:
+        sys.exit(1)
```
Comment (apahim) [inline reconcile/kafka_clusters.py:116]: use reconcile.status.ExitCodes.ERROR pls

File: reconcile/kafka_clusters.py (hunk 1 above)
Comment (apahim) [reply reconcile/kafka_clusters.py:50]: (and the corresponding `kafkaCluster`)

File: reconcile/kafka_clusters.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        desired_cluster = [c for c in desired_state
+                           if kafka_cluster_name == c['name']][0]
+        current_cluster = [c for c in current_state
+                           if kafka_cluster_name == c['name']]
+        # check if cluster exists. if not - create it
+        if not current_cluster:
+            logging.info(['create_cluster', kafka_cluster_name])
+            if not dry_run:
+                ocm = ocm_map.get(kafka_cluster_name)
+                ocm.create_kafka_cluster(desired_cluster)
```
Comment (apahim) [inline reconcile/kafka_clusters.py:81]: We probably have to do error handling here...
```
[2020-11-11 16:54:32] [DEBUG] [connectionpool.py:_make_request:391] - https://api.stage.openshift.com:443 "POST /api/managed-se
rvices-api/v1/kafkas?async=true HTTP/1.1" 400 203
[2020-11-11 16:54:32] [ERROR] [ocm.py:_post:505] - {"id":"8","kind":"Error","href":"/api/managed-services-api/v1/errors/8","cod
e":"MGD-SERV-API-8","reason":"region us-west-1 is not supported for aws, supported regions are: [us-east-1]","operation_id":"1k
9XEtrscyNpdPXjMesFYc1dRRN"}
Traceback (most recent call last):
  File "/home/apahim/git/qontract-reconcile/venv/bin/qontract-reconcile", line 33, in <module>
    sys.exit(load_entry_point('reconcile', 'console_scripts', 'qontract-reconcile')())
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/utils/binary.py", line 14, in f_binary
    f(*args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/reconcile/cli.py", line 1100, in kafka_clusters
    run_integration(reconcile.kafka_clusters, ctx.obj, thread_pool_size,
  File "/home/apahim/git/qontract-reconcile/reconcile/cli.py", line 335, in run_integration
    func_container.run(dry_run, *args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/utils/defer.py", line 15, in func_wrapper
    return func(*args, defer=defer, **kwargs)
  File "/home/apahim/git/qontract-reconcile/reconcile/kafka_clusters.py", line 81, in run
    ocm.create_kafka_cluster(desired_cluster)
  File "/home/apahim/git/qontract-reconcile/utils/ocm.py", line 492, in create_kafka_cluster
    self._post(api, data)
  File "/home/apahim/git/qontract-reconcile/utils/ocm.py", line 506, in _post
    raise e
  File "/home/apahim/git/qontract-reconcile/utils/ocm.py", line 503, in _post
    r.raise_for_status()
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/requests/models.py", line 940, in raise_for_stat
us
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.stage.openshift.com/api/managed-services-api/
v1/kafkas?async=true
```

**Review by apahim — CHANGES_REQUESTED** [review]
Great work with kafka clusters!
We need a couple of improvements related to exception handling.

**Comment by apahim** [comment]
In case you're wondering...
kafka-clusters

File: reconcile/kafka_clusters.py (hunk 1 above)
Comment (maorfr) [reply reconcile/kafka_clusters.py:55]: added an exit(0) if no kafka clusters are defined

File: reconcile/kafka_clusters.py (hunk 2 above)
Comment (maorfr) [reply reconcile/kafka_clusters.py]: thanks!

File: reconcile/kafka_clusters.py (hunk 3 above)
Comment (maorfr) [reply reconcile/kafka_clusters.py:86]: i've added a region validation in https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/11847.
i don't want to handle errors yet, i want us to see all the errors for this new integration (and service). in the future, i believe they will add preflight checks that we will be able to use as a preventive measure. wdyt?

Comment (apahim) [reply reconcile/kafka_clusters.py:89]: soungs like a plan :)

**Review by apahim — APPROVED** [review]
Ship it! :)

**Pull request merged**